### PR TITLE
[Draft] Multi-node Stable Diffusion 

### PR DESCRIPTION
**Pull request by sean-smith**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    cmake \
+    apt-utils \
+    devscripts \
+    debhelper \
+    libsubunit-dev \
+    check \
+    pkg-config
+
+RUN mkdir -p /var/run/sshd
+RUN sed -i 's/[ #]\(.*StrictHostKeyChecking \).*/ \1no/g' /etc/ssh/ssh_config && \
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile:34]: required?

File: 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    echo "    UserKnownHostsFile /dev/null" >> /etc/ssh/ssh_config && \
+    sed -i 's/#\(StrictModes \).*/\1no/g' /etc/ssh/sshd_config
+ENV LD_LIBRARY_PATH /usr/local/cuda/extras/CUPTI/lib64:/opt/amazon/openmpi/lib:/opt/nccl/build/lib:/opt/amazon/efa/lib:/opt/aws-ofi-nccl/install/lib:/usr/local/lib:$LD_LIBRARY_PATH
+ENV PATH /opt/amazon/openmpi/bin/:/opt/amazon/efa/bin:/usr/bin:/usr/local/bin:$PATH
+RUN curl https://bootstrap.pypa.io/get-pip.py -o /tmp/get-pip.py \
+ && python3 /tmp/get-pip.py \
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile:40]: indentation is not consistent

File: 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    && git checkout ${NCCL_TESTS_VERSION} \
+    && make MPI=1 \
+    MPI_HOME=/opt/amazon/openmpi/ \
+    CUDA_HOME=/usr/local/cuda \
+    NCCL_HOME=/opt/nccl/build \
+    NVCC_GENCODE="-gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_90,code=sm_90"
+
+RUN git clone https://github.com/mosaicml/diffusion-benchmark.git
+RUN pip3 install -r diffusion-benchmark/requirements.txt
+RUN pip3 install mosaicml==0.15.0 --force
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile:96]: version as an env var?

File: 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    #     OPTION="exclude"
+    # fi
+    srun --nodelist=${NODE} --ntasks=1 -l "${ARGS[@]}" composer \
+        --world_size ${WORLD_SIZE} \
+        --nproc ${NPROC} \
+        --node_rank ${NODE_RANK} \
+        --master_addr ${MASTER_ADDR} \
+        --master_port ${MASTER_PORT} \
+        --verbose \
+        benchmark.py \
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch:60]: where is this file?

**Review by perifaws — CHANGES_REQUESTED** [review]

File: 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile (hunk 1 above)
Comment (awsankur) [reply 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile:34]: These lines are not specific to the stable diffusion model. Taken from the Dockerfile for NCCL tests:
https://github.com/aws-samples/awsome-distributed-training/blob/multi-node/4.validation_scripts/0.nccl-tests/0.nccl-tests.Dockerfile

File: 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile (hunk 3 above)
Comment (awsankur) [reply 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile:96]: Added new env variable

File: 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch (hunk 4 above)
Comment (awsankur) [reply 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch:60]: The multi-node Dockerfile is cloning the MosaicML stable diffusion benchmark repo and `benchmark.py` lies in the WORKDIR `/workspace/diffusion-benchmark`

File: 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        --use_ema --use_synth_data --device_train_microbatch_size 4
+}
+
+NODE_RANK=1
```
[the remark this answers is not in the document]
Comment (awsankur) [reply 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch:64]: Yes, MosaicML's stable diffusion is a pretty popular implementation. This workaround to get MosaicML's Composer working with Slurm might benefit from a review by MosaicML.

File: 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+RUN pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu121 --force
+RUN pip3 uninstall transformer-engine -y
+RUN pip3 install protobuf==3.20.3
+
+RUN rm -rf /var/lib/apt/lists/*
+ENV LD_PRELOAD /opt/nccl/build/lib/libnccl.so
```
[the remark this answers is not in the document]
Comment (awsankur) [reply 3.test_cases/6.stable-diffusion/multi-node/1.Dockerfile:102]: Not 100% sure. Probably safe to keep it.

File: 3.test_cases/6.stable-diffusion/single-node/0.Dockerfile [hunk 7]
```diff
@@ -1,12 +1,17 @@
 # Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 # SPDX-License-Identifier: MIT-0
 
-FROM nvcr.io/nvidia/pytorch:23.08-py3
+ARG PYTORCH_IMAGE
+
+FROM PYTORCH_IMAGE
+
+ARG MOSAICML_VERSION
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/single-node/0.Dockerfile:8]: Use default values

File: 3.test_cases/6.stable-diffusion/README.md [hunk 8]
```diff
@@ -1,18 +1,27 @@
...
+export DOCKER_IMAGE_NAME=mosaicml-stable-diffusion
+export TAG=$MOSAICML_VERSION
+```
 
-Just for simplifaction of testing, we have separate scripts for Single node and Multi node Distributed Training. We will also present a comparison of throughput (images/second) achieved with P4de (A100 80GB) and P5 (H100 80GB) instances.
 
-## 0. Conda and Docker
+## 0. Create Conda Environment
 
 Make sure you are able to create conda environments and docker containers. For example, to install Miniconda, please follow the steps below:
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/README.md:26]: You may want to clarify why one need Conda while we are using containers

File: 3.test_cases/6.stable-diffusion/README.md [hunk 9]
```diff
@@ -68,15 +95,13 @@ Once this change is done, you can install composer as `pip3 install -e .`
 The `single-node` folder also has the Dockerfile and a Makefile with commands to build the image and run the container.
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/README.md:95]: Are there thing to know about his container? What does it do?

File: 3.test_cases/6.stable-diffusion/README.md [hunk 10]
```diff
@@ -68,15 +95,13 @@ Once this change is done, you can install composer as `pip3 install -e .`
...
+# run container
+docker run --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 -it ${DOCKER_IMAGE_NAME} /bin/bash
 
-# run it
-docker run --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 -it mosaicml-stable-diffusion /bin/bash
-# or you can do
-# make run
 ```
 
 ### 1.2 Single Node Training
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/README.md:107]: Same title as the section?

File: 3.test_cases/6.stable-diffusion/README.md [hunk 11]
```diff
@@ -142,8 +167,45 @@ from diffusers.models.attention_processor import AttnProcessor
 self.unet = UNet2DConditionModel.from_pretrained(''stabilityai/stable-diffusion-2-base'', subfolder='unet')
 self.unet.set_attn_processor(AttnProcessor())
 ```
-AttnProcessor2_0 which is a Processor for implementing scaled dot-product attention is enabled by default if you're using PyTorch 2.0.
+`AttnProcessor2_0` which is a Processor for implementing scaled dot-product attention is enabled by default if you're using PyTorch 2.0.
 
-The older self.unet.set_attn_processor(AttnProcessor()) gives Cuda OOM error with a batch size of 32 while with `AttnProcessor2_0()` is able to run with a batch size of 32 and yield 385 images/sec throughput 
+The older `self.unet.set_attn_processor(AttnProcessor())` gives Cuda OOM error with a batch size of 32 while with `AttnProcessor2_0()` is able to run with a batch size of 32 and yield 385 images/sec throughput
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/README.md:172]: Could we reference to a CUDA OOM? (just a link)

File: 3.test_cases/6.stable-diffusion/README.md [hunk 12]
```diff
@@ -142,8 +167,45 @@ from diffusers.models.attention_processor import AttnProcessor
...
-The older self.unet.set_attn_processor(AttnProcessor()) gives Cuda OOM error with a batch size of 32 while with `AttnProcessor2_0()` is able to run with a batch size of 32 and yield 385 images/sec throughput 
+The older `self.unet.set_attn_processor(AttnProcessor())` gives Cuda OOM error with a batch size of 32 while with `AttnProcessor2_0()` is able to run with a batch size of 32 and yield 385 images/sec throughput 
 
 More details on this can be found here: https://pytorch.org/blog/accelerated-diffusers-pt-20/
+
+## 2. Multi Node Tests
+
+### 2.1 Multi-Node Training
+
+For the multi-node training we've created a `Dockerfile`, and Slurm submit script and a `Makefile` to build the docker image and convert it to an enroot image. To get started please follow the guide [AWS ParallelCluster Distributed Training](../../1.architectures/2.aws-parallelcluster). Before starting this section make sure you have the following setup:
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/README.md:180]: remove the makefile reference

File: 3.test_cases/6.stable-diffusion/README.md [hunk 13]
```diff
@@ -142,8 +167,45 @@ from diffusers.models.attention_processor import AttnProcessor
...
+* FSx Lustre Filesystem
+
+1. To get started, clone this repo and cd into the multi-node directory:
+
+```
+git clone https://github.com/aws-samples/awsome-distributed-training.git
+cd awsome-distributed-training/6.stable-diffusion/multi-node
+```
+
+Next build the docker image:
```
Comment (perifaws) [inline 3.test_cases/6.stable-diffusion/README.md:194]: no more numbered list?

File: 3.test_cases/6.stable-diffusion/README.md (hunk 8 above)
Comment (awsankur) [reply 3.test_cases/6.stable-diffusion/README.md:26]: Added clarification

File: 3.test_cases/6.stable-diffusion/README.md [hunk 14]
```diff
@@ -1,27 +1,44 @@
...
 ``` 
 
+To simplify testing, we have separate scripts for Single node and Multi node Distributed Training. We will also present a comparison of throughput (images/second) achieved with P4de (A100 80GB) and P5 (H100 80GB) instances.
+
+You can export the following environment variables:
+
+```
+export PYTHON_VERSION=3.10
+# We are using Python version 3.10 in this work. For a different Python version select the right Miniconda file from https://repo.anaconda.com/miniconda/
+export MINICONDA_INSTALLER=Miniconda3-py310_23.5.2-0-Linux-x86_64
```
Comment (KeitaW) [inline 3.test_cases/6.stable-diffusion/README.md:17]: If `miniconda` example is deprecated, this variable can be removed too.

File: 3.test_cases/6.stable-diffusion/README.md [hunk 15]
```diff
@@ -35,15 +52,32 @@ elif [[ "$os" == "ubuntu" ]]; then
...
+
+
 ```
 
-## 1 Single Node Training
 
-### 1.1 Single Node Setup
+### 1.2 Single Node Setup with Conda
 
 When testing the latest version of MosaicML's Composer, we found that different ways to set up the environment with a PyTorch Nightly conda environment or a Nvidia PyTorch Docker container. For single or multi-node testing, you could use either to run distributed training. Next, we present both approaches.
```
Comment (KeitaW) [inline 3.test_cases/6.stable-diffusion/README.md:78]: This sentence is also need to be revised.

File: 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+#SBATCH --exclusive # job has exclusive use of the resource, no sharing
+#SBATCH --wait-all-nodes=1
+#SBATCH --output jobs/slurm-%j.out
+
+# default variables for Enroot
+: "${APPS_PATH:=/fsx}"
+: "${DATA_PATH:=/fsx}"
+
+# default variables for Enroot
+: "${IMAGE:=$APPS_PATH/mosaicml/stable-diffusion.sqsh}"
```
Comment (KeitaW) [inline 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch:20]: According to the  `README.md` the sqsh file supposed to be stored as `$APPS_PATH/mosaicml-stable-diffusion.sqsh `.

File: 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+#SBATCH --job-name=mosaicml-stable-diffusion # name of your job
+#SBATCH --gpus-per-node=8 # Number of GPU per node
+#SBATCH --gres=gpu:8 # number of GPU we reserve
+#SBATCH --gpus-per-task=8 # Number of GPU per node
+#SBATCH --exclusive # job has exclusive use of the resource, no sharing
+#SBATCH --wait-all-nodes=1
+#SBATCH --output jobs/slurm-%j.out
+
+# default variables for Enroot
+: "${APPS_PATH:=/fsx}"
```
Comment (KeitaW) [inline 3.test_cases/6.stable-diffusion/multi-node/2.train.sbatch:16]: In `RERADME.md`, `APPS_PATH` is implicitly specified as `/apps`. Better to follow to the default value here.

File: 3.test_cases/6.stable-diffusion/single-node/0.Dockerfile [hunk 18]
```diff
@@ -1,12 +1,17 @@
...
+ARG PYTORCH_INDEX_URL
 
 RUN git clone https://github.com/mosaicml/diffusion-benchmark.git /wd
 RUN pip3 install -r /wd/requirements.txt
-RUN pip3 install mosaicml==0.15.0 --force
-RUN pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu121 --force
+RUN pip3 install mosaicml==${MOSAICML_VERSION} --force
+RUN pip3 install --pre torch torchvision torchaudio --index-url ${PYTORCH_INDEX_URL} --force
 RUN pip3 uninstall transformer-engine -y
```
Comment (KeitaW) [inline 3.test_cases/6.stable-diffusion/single-node/0.Dockerfile:15]: I am not sure if the current example making use of transformer-engine. It seems that the original repo isn't integrate the feature yet?

**Review by KeitaW — COMMENTED** [review]
Left some minor comments. Kudos @sean-smith  and @awsankur  for this great work!

**Comment by KeitaW** [comment]
[This is a general comment. No need to address in this PR.] I believe that the only difference between the docker image of `single-node` and `multi-node` is EFA/NCCL related settings. We may use multi-stage build feature to unify those two.

**Comment by KeitaW** [comment]
Hmm... actually my multi-node job failed with the following error on P5 cluster.
```
0: p5-st-p5-2:947575:949672 [0] NCCL INFO proxy.cc:858 -> 2 [Proxy Thread]
0: wandb: Tracking run with wandb version 0.13.11
0: wandb: W&B syncing is set to `offline` in this directory.  
0: wandb: Run `wandb online` or set WANDB_MODE=online to enable cloud syncing.
0: wandb: WARNING URL not available in offline run
0: p5-st-p5-1:1130982:1133081 [0] NCCL INFO [Service thread] Connection closed by localRank 4
0: p5-st-p5-1:1130982:1133081 [0] NCCL INFO [Service thread] Connection closed by localRank 2
0: p5-st-p5-1:1130982:1133081 [0] NCCL INFO [Service thread] Connection closed by localRank 3
```
any thoughts on the root cause? I'm working with 
* pclustre 3.7.2
* AMI: Base DLAMI 10/26

**Comment by KeitaW** [comment]
Hmm... although the job is up and running, I don't see any meaningful output. Any thoughts on the cause? I'm using 
* Pcluster 3.7.2
* AMI: Base DLAMI 10/16

```
$ squeue
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
               277        p5 mosaicml   ubuntu  R      53:51      2 p5-st-p5-[1-2]

$ tail -f jobs/slurm-277.out 
0: p5-st-p5-2:947575:949672 [0] NCCL INFO transport/net.cc:1112 -> 2
0: p5-st-p5-2:947575:949672 [0] NCCL INFO proxy.cc:679 -> 2
0: p5-st-p5-2:947575:949672 [0] NCCL INFO proxy.cc:858 -> 2 [Proxy Thread]
0: wandb: Tracking run with wandb version 0.13.11
0: wandb: W&B syncing is set to `offline` in this directory.  
0: wandb: Run `wandb online` or set WANDB_MODE=online to enable cloud syncing.
0: wandb: WARNING URL not available in offline run
0: p5-st-p5-1:1130982:1133081 [0] NCCL INFO [Service thread] Connection closed by localRank 4
0: p5-st-p5-1:1130982:1133081 [0] NCCL INFO [Service thread] Connection closed by localRank 2
0: p5-st-p5-1:1130982:1133081 [0] NCCL INFO [Service thread] Connection closed by localRank 3
```

File: 3.test_cases/6.stable-diffusion/README.md (hunk 15 above)
Comment (awsankur) [reply 3.test_cases/6.stable-diffusion/README.md:78]: For Single Node Section 1.2 shows setup with a Conda environment and 1.3 an optional step with Docker.

File: 3.test_cases/6.stable-diffusion/single-node/0.Dockerfile (hunk 18 above)
Comment (awsankur) [reply 3.test_cases/6.stable-diffusion/single-node/0.Dockerfile:15]: Yes

**Comment by awsankur** [comment]
> Hmm... although the job is up and running, I don't see any meaningful output. Any thoughts on the cause? I'm using
> 
> * Pcluster 3.7.2
> * AMI: Base DLAMI 10/16
> 
> ```
> $ squeue
>              JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
>                277        p5 mosaicml   ubuntu  R      53:51      2 p5-st-p5-[1-2]
> 
> $ tail -f jobs/slurm-277.out 
> 0: p5-st-p5-2:947575:949672 [0] NCCL INFO transport/net.cc:1112 -> 2
> 0: p5-st-p5-2:947575:949672 [0] NCCL INFO proxy.cc:679 -> 2
> 0: p5-st-p5-2:947575:949672 [0] NCCL INFO proxy.cc:858 -> 2 [Proxy Thread]
> 0: wandb: Tracking run with wandb version 0.13.11
> 0: wandb: W&B syncing is set to `offline` in this directory.  
> 0: wandb: Run `wandb online` or set WANDB_MODE=online to enable cloud syncing.
> 0: wandb: WARNING URL not available in offline run
> 0: p5-st-p5-1:1130982:1133081 [0] NCCL INFO [Service thread] Connection closed by localRank 4
> 0: p5-st-p5-1:1130982:1133081 [0] NCCL INFO [Service thread] Connection closed by localRank 2
> 0: p5-st-p5-1:1130982:1133081 [0] NCCL INFO [Service thread] Connection closed by localRank 3
> ```

I have made changes to the Dockerfile to make sure this issue does not happen and training proceeds successfully.

**Pull request merged**